crates/ra_hir/src/semantics.rs

```diff
@@ -104,6 +104,13 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         tree
     }
 
+    pub fn ast<T: AstNode>(&self, ast_ptr: &AstPtr<T>, file_id: HirFileId) -> T {
```

Comment:
hmm I think part of the point of `Semantics` is not having to deal with `HirFileId`s. maybe this could instead directly take an `impl AstDiagnostic`?

Reply:
`Semantics::cache` requires a `root_node` and ` HirFileId` though.
I could make the argument a `Diagnostic + AstDiagnostic` and use `Diagnostic::source` to get the `HirFileId`... Would that be better?
There would be 2 `AstDatabase::parse_or_expand()` calls, but that's cached by Salsa so it should be fine?

Reply:
Sorry, yes, that was indeed what I was thinking of.